prompt2model/param_selector/search_with_optuna.py

```diff
@@ -0,0 +1,239 @@
+            validation_sets (Dataset): A dataset for computing validation metrics.
+            hyperparameter_space (Optional[dict[str, Any]], optional): The set
+                of possible values of hyperparaneters values required for doing
+                optimal hyperparameter search. Defaults to None.
+
+        Returns:
+            Returns a dict which contains the best hyperparameters
+        """
+        supported_hp_space_keys = set(DEFAULT_HYPERPARAMETERS_SPACE.keys())
+        if hyperparameters is not None:
+            assert set(hyperparameters.keys()).issubset(
+                supported_hp_space_keys
+            ), f"Only support {supported_hp_space_keys} as training parameters."
+        hyperparameter_space = self._build_hp_space(hyperparameters)
+
+        concatenated_training_dataset = concatenate_datasets(training_datasets)
+        train_dataset = self.generation_model_trainer.tokenize_dataset(
+            concatenated_training_dataset
+        )
+
+        def objective(trial: Trial):
+            model = self.generation_model_trainer.model
+            training_args = transformers.TrainingArguments(
+                output_dir="./checkpoint",
+                learning_rate=trial.suggest_loguniform(
+                    "learning_rate",
+                    low=hyperparameter_space["min_learning_rate"],
+                    high=hyperparameter_space["max_learning_rate"],
+                ),
+                weight_decay=trial.suggest_loguniform(
+                    "weight_decay",
+                    low=hyperparameter_space["min_weight_decay"],
+                    high=hyperparameter_space["max_weight_decay"],
+                ),
+                num_train_epochs=trial.suggest_int(
+                    "num_train_epochs",
+                    low=hyperparameter_space["min_num_train_epochs"],
+                    high=hyperparameter_space["max_num_train_epochs"],
+                ),
+                save_strategy=trial.suggest_categorical(
+                    "save_strategy", hyperparameter_space["save_strategy"]
+                ),
+                per_device_train_batch_size=trial.suggest_categorical(
+                    "per_device_train_batch_size",
+                    hyperparameter_space["per_device_train_batch_size"],
+                ),
+                per_device_eval_batch_size=hyperparameter_space[
+                    "per_device_train_batch_size"
+                ],
+            )
+            objective_trainer = transformers.Trainer(
+                model=model,
+                args=training_args,
+                data_collator=transformers.DataCollatorForSeq2Seq(
+                    tokenizer=self.generation_model_trainer.tokenizer
+                ),
+                train_dataset=train_dataset,
+                eval_dataset=validation,
+            )
+
+            result = objective_trainer.train()
+            return result.training_loss
```

Comment:
We should be selecting hyperparameters based on validation metrics (e.g. validation loss) rather than training metrics

Reply:
As discussed, we can move this to a different PR right?